File: cpp/src/level.cpp
// Hand-authored map.
//
//   alcove (2, green, low ceiling)
//        v10──v11
//         │    │
//         │    │                      east (1, red, raised + tall)
//  v0──v1─┴────┴─v2──v3────v4    v4──v12
//   │                      │    │    │
//   │        main (0, gray)│door│    │
//   │                      │    │    │
//   │                      │    │    │
//   │                      │v5──v13
//   │                      │
//  v9──v8────v7──v6
//       │     \
//       │      \       corridor (3, blue, slight step up + drop ceiling)
//       │       \
//      v15──────v14
//       │        │     south (4, violet, sunken floor + tall ceiling)
//      v17──────v16     -- trapezoid with angled east/west walls

#include "level.hpp"

const std::vector<Vec2> vertices = {
    // Main hall perimeter
    {  0,    0},  //  0 main NW
    { 80,    0},  //  1 alcove opening west
    {140,    0},  //  2 alcove opening east
    {240,    0},  //  3 main NE
    {240,   80},  //  4 door top
    {240,  140},  //  5 door bottom
    {240,  200},  //  6 main SE
    {150,  200},  //  7 corridor opening east
    { 90,  200},  //  8 corridor opening west
    {  0,  200},  //  9 main SW
    // Alcove
    { 80,  -50},  // 10 alcove NW
    {140,  -50},  // 11 alcove NE
    // East room
    {360,   80},  // 12 east NE
    {360,  140},  // 13 east SE
    // Corridor
    {200,  300},  // 14 corridor SE / south NE
    {140,  300},  // 15 corridor SW / south NW
    // South chamber
    {240,  360},  // 16 south SE
    {100,  360},  // 17 south SW
};

// Per-sector wall colors, chosen so each sector reads as its own palette.
namespace {
    constexpr RGBA mainWall   = {158, 144, 115, 255};
    constexpr RGBA mainUpper  = {110, 105,  92, 255};
    constexpr RGBA mainLower  = {118,  96,  72, 255};
    constexpr RGBA eastWall   = {196,  90,  62, 255};
    constexpr RGBA alcoveWall = {108, 170,  82, 255};
    constexpr RGBA corrWall   = { 94, 134, 200, 255};
    constexpr RGBA southWall  = {168, 108, 206, 255};
}

const std::vector<Sector> sectors = {
    // 0: Main hall — warm tan
    {  0,  64, {82,  76, 60, 255}, {50, 56, 70, 255}, 0.85},
    // 1: East room — red, raised floor, taller ceiling
    { 16,  88, {112, 62, 46, 255}, {72, 44, 34, 255}, 0.95},
    // 2: Alcove — green, low ceiling, raised floor
    {-12,  48, {58,  84, 52, 255}, {42, 62, 40, 255}, 0.9},
    // 3: Diagonal corridor — blue, slight step up
    {  4,  68, {48,  62, 96, 255}, {34, 46, 78, 255}, 0.82},
    // 4: South chamber — violet, sunken floor
    {-12,  52, {76,  52, 96, 255}, {52, 38, 76, 255}, 0.7},
};

const std::vector<LineDef> linedefs = {
    // ---- Main hall perimeter (front = 0) ----
    {0, 1, 0, noSector, mainWall, mainUpper, mainLower},
    {1, 2, 0, 2,        mainWall, alcoveWall, mainLower},
    {2, 3, 0, noSector, mainWall, mainUpper, mainLower},
    {3, 4, 0, noSector, mainWall, mainUpper, mainLower},
    {4, 5, 0, 1,        mainWall, mainUpper, eastWall},
    {5, 6, 0, noSector, mainWall, mainUpper, mainLower},
    {6, 7, 0, noSector, mainWall, mainUpper, mainLower},
    {7, 8, 0, 3,        mainWall, corrWall, corrWall},
    {8, 9, 0, noSector, mainWall, mainUpper, mainLower},
    {9, 0, 0, noSector, mainWall, mainUpper, mainLower},

    // ---- Alcove (front = 2) ----
    { 1, 10, 2, noSector, alcoveWall, alcoveWall, alcoveWall},
    {10, 11, 2, noSector, alcoveWall, alcoveWall, alcoveWall},
    {11,  2, 2, noSector, alcoveWall, alcoveWall, alcoveWall},

    // ---- East room (front = 1) ----
    { 4, 12, 1, noSector, eastWall, eastWall, eastWall},
    {12, 13, 1, noSector, eastWall, eastWall, eastWall},
    {13,  5, 1, noSector, eastWall, eastWall, eastWall},

    // ---- Diagonal corridor (front = 3) ----
    { 7, 14, 3, noSector, corrWall, corrWall, corrWall},
    {14, 15, 3, 4,        corrWall, corrWall, southWall},
    {15,  8, 3, noSector, corrWall, corrWall, corrWall},

    // ---- South chamber (front = 4) ----
    {14, 16, 4, noSector, southWall, southWall, southWall},
    {16, 17, 4, noSector, southWall, southWall, southWall},
    {17, 15, 4, noSector, southWall, southWall, southWall},
};

File: go/level.go
package main

// Hand-authored map. See the ASCII sketch in the original Level.swift.

var vertices = []Vec2{
	// Main hall perimeter
	{0, 0},     // 0  main NW
	{80, 0},    // 1  alcove opening west
	{140, 0},   // 2  alcove opening east
	{240, 0},   // 3  main NE
	{240, 80},  // 4  door top
	{240, 140}, // 5  door bottom
	{240, 200}, // 6  main SE
	{150, 200}, // 7  corridor opening east
	{90, 200},  // 8  corridor opening west
	{0, 200},   // 9  main SW
	// Alcove
	{80, -50},  // 10 alcove NW
	{140, -50}, // 11 alcove NE
	// East room
	{360, 80},  // 12 east NE
	{360, 140}, // 13 east SE
	// Corridor
	{200, 300}, // 14 corridor SE / south NE
	{140, 300}, // 15 corridor SW / south NW
	// South chamber
	{240, 360}, // 16 south SE
	{100, 360}, // 17 south SW
}

var (
	mainWall   = RGBA{158, 144, 115, 255}
	mainUpper  = RGBA{110, 105, 92, 255}
	mainLower  = RGBA{118, 96, 72, 255}
	eastWall   = RGBA{196, 90, 62, 255}
	alcoveWall = RGBA{108, 170, 82, 255}
	corrWall   = RGBA{94, 134, 200, 255}
	southWall  = RGBA{168, 108, 206, 255}
)

var sectors = []Sector{
	// 0: Main hall — warm tan
	{floorH: 0, ceilH: 64,
		floorColor: RGBA{82, 76, 60, 255},
		ceilColor:  RGBA{50, 56, 70, 255},
		light:      0.85},
	// 1: East room — red, raised floor, taller ceiling
	{floorH: 16, ceilH: 88,
		floorColor: RGBA{112, 62, 46, 255},
		ceilColor:  RGBA{72, 44, 34, 255},
		light:      0.95},
	// 2: Alcove — green, low ceiling, raised floor
	{floorH: -12, ceilH: 48,
		floorColor: RGBA{58, 84, 52, 255},
		ceilColor:  RGBA{42, 62, 40, 255},
		light:      0.9},
	// 3: Diagonal corridor — blue, slight step up
	{floorH: 4, ceilH: 68,
		floorColor: RGBA{48, 62, 96, 255},
		ceilColor:  RGBA{34, 46, 78, 255},
		light:      0.82},
	// 4: South chamber — violet, sunken floor
	{floorH: -12, ceilH: 52,
		floorColor: RGBA{76, 52, 96, 255},
		ceilColor:  RGBA{52, 38, 76, 255},
		light:      0.7},
}

var linedefs = []LineDef{
	// ---- Main hall perimeter (front = 0) ----
	{v1: 0, v2: 1, frontSector: 0, backSector: noSector, wallColor: mainWall, upperColor: mainUpper, lowerColor: mainLower},
	{v1: 1, v2: 2, frontSector: 0, backSector: 2, wallColor: mainWall, upperColor: alcoveWall, lowerColor: mainLower},
	{v1: 2, v2: 3, frontSector: 0, backSector: noSector, wallColor: mainWall, upperColor: mainUpper, lowerColor: mainLower},
	{v1: 3, v2: 4, frontSector: 0, backSector: noSector, wallColor: mainWall, upperColor: mainUpper, lowerColor: mainLower},
	{v1: 4, v2: 5, frontSector: 0, backSector: 1, wallColor: mainWall, upperColor: mainUpper, lowerColor: eastWall},
	{v1: 5, v2: 6, frontSector: 0, backSector: noSector, wallColor: mainWall, upperColor: mainUpper, lowerColor: mainLower},
	{v1: 6, v2: 7, frontSector: 0, backSector: noSector, wallColor: mainWall, upperColor: mainUpper, lowerColor: mainLower},
	{v1: 7, v2: 8, frontSector: 0, backSector: 3, wallColor: mainWall, upperColor: corrWall, lowerColor: corrWall},
	{v1: 8, v2: 9, frontSector: 0, backSector: noSector, wallColor: mainWall, upperColor: mainUpper, lowerColor: mainLower},
	{v1: 9, v2: 0, frontSector: 0, backSector: noSector, wallColor: mainWall, upperColor: mainUpper, lowerColor: mainLower},

	// ---- Alcove (front = 2) ----
	{v1: 1, v2: 10, frontSector: 2, backSector: noSector, wallColor: alcoveWall, upperColor: alcoveWall, lowerColor: alcoveWall},
	{v1: 10, v2: 11, frontSector: 2, backSector: noSector, wallColor: alcoveWall, upperColor: alcoveWall, lowerColor: alcoveWall},
	{v1: 11, v2: 2, frontSector: 2, backSector: noSector, wallColor: alcoveWall, upperColor: alcoveWall, lowerColor: alcoveWall},

	// ---- East room (front = 1) ----
	{v1: 4, v2: 12, frontSector: 1, backSector: noSector, wallColor: eastWall, upperColor: eastWall, lowerColor: eastWall},
	{v1: 12, v2: 13, frontSector: 1, backSector: noSector, wallColor: eastWall, upperColor: eastWall, lowerColor: eastWall},
	{v1: 13, v2: 5, frontSector: 1, backSector: noSector, wallColor: eastWall, upperColor: eastWall, lowerColor: eastWall},

	// ---- Diagonal corridor (front = 3) ----
	{v1: 7, v2: 14, frontSector: 3, backSector: noSector, wallColor: corrWall, upperColor: corrWall, lowerColor: corrWall},
	{v1: 14, v2: 15, frontSector: 3, backSector: 4, wallColor: corrWall, upperColor: corrWall, lowerColor: southWall},
	{v1: 15, v2: 8, frontSector: 3, backSector: noSector, wallColor: corrWall, upperColor: corrWall, lowerColor: corrWall},

	// ---- South chamber (front = 4) ----
	{v1: 14, v2: 16, frontSector: 4, backSector: noSector, wallColor: southWall, upperColor: southWall, lowerColor: southWall},
	{v1: 16, v2: 17, frontSector: 4, backSector: noSector, wallColor: southWall, upperColor: southWall, lowerColor: southWall},
	{v1: 17, v2: 15, frontSector: 4, backSector: noSector, wallColor: southWall, upperColor: southWall, lowerColor: southWall},
}
